abci/src/application.rs

```diff
@@ -98,6 +98,14 @@ pub trait Application: Send + Clone + 'static {
     ) -> ResponseApplySnapshotChunk {
         Default::default()
     }
+
+    fn prepare_proposal(&self, _request: RequestPrepareProposal) -> ResponsePrepareProposal {
+        Default::default()
+    }
+
+    fn process_proposal(&self, _request: RequestProcessProposal) -> ResponseProcessProposal {
+        Default::default()
+    }
```

Comment:
Just to note that the defaults will probably not work (they will propose empty blocks and think [reject](https://github.com/informalsystems/tendermint-rs/pull/1193/files#diff-f98b59e5344ef9d1c053ec7d1dd516b0627d1e390a9ba8813ab004c70b340c9fR15) all proposals). The spec has [instructions](https://github.com/tendermint/tendermint/blob/v0.37.0-rc2/spec/abci/abci%2B%2B_tmint_expected_behavior.md#adapting-existing-applications-that-use-abci) for adaptation.

Reply:
Right; sorry about this, `tendermint-abci` has not been my top priority for updating to 0.37. Will fix.

Reply:
As I understand it, this whole default implementation only provides the most trivial default responses.
It's up to the application to fill the methods in where necessary.

Reply:
Yes, I agree; what I'm trying to say is that by leaving it on defaults, filling in these methods will always be necessary. 

I can't vouch for all the other cases, but I would expect for those methods the default values represent an acceptable behaviour, like `check_tx` not finding any error, or `end_block` not changing the power table. 

But with these two new methods, the defaults are not accepted by Tendermint:
* If we return default from `prepare_proposal`, that effectively will remove all transactions Tendermint wanted to propose, resulting in an empty block instead. 
* If then we return `ProcessProposal::Unknown` (which is the default, although it could be `Accept`) instead of `ProcessProposal::Accept`, we get the following error: `CONSENSUS FAILURE!!!                         module=consensus err="ProcessProposal responded with status UNKNOWN"`. That is because Tendermint now considers us insane: we just asked it to propose a block we ourselves do not accept.

So, if we want to have a transition to v0.37 where users of the `Application` trait don't have to override unless they have something to change from the default behaviour, and they expect Tendermint to just work, then the default method implementations have a bit more work to do than usual.

(NB the specs explicitly say that TC _may_ pass more transactions in `PrepareProposal` than what fits into the maximum size; otherwise the implementation would be a trivial copy).

Reply:
@aakoshh Would that work for you or would you rather leave out the default impls?

Reply:
Thanks for asking @romac, my vote goes for having a default implementation that copies transactions and accepts proposals, unless overridden, like @mzabaluev said. This is so people don't have to come up with boilerplate to do this every time they implement the trait, similar to how they don't have to override `check_tx` unless they want to actually perform some logic. 

For example I forked `tower-abci` to point it at this branch of `tendermint-rs` and had to implement this logic [for the kvstore example](https://github.com/consensus-shipyard/tower-abci/pull/1/files#diff-3064912e722a6c3a27e57368a923034ba4dff6408ddfc4923ef4e00ed68b38bc), where it seems a bit accidental among all the other `Default::default()`s. After that I took inspiration from the `Application` trait in `tendermint-rs` and provided an [async version](https://github.com/consensus-shipyard/fendermint/blob/master/fendermint/abci/src/application.rs) which takes care of these defaults, so I think it's the sensible thing to do.

Reply:
Thank you @aakoshh for your feedback.
The more elaborate (but still very naive) default behaviours have now been implemented in https://github.com/informalsystems/tendermint-rs/pull/1193/commits/7e79039558b50fcdd3ca47512b4e58ed6624132f